File: apps/api/models/delays.go
package models

import "time"

// ServiceAlert represents a transit service alert
type ServiceAlert struct {
	AlertID           string   `json:"alertId"`
	Cause             string   `json:"cause,omitempty"`
	Effect            string   `json:"effect,omitempty"`
	DescriptionText   string   `json:"descriptionText"`
	AffectedRoutes    []string `json:"affectedRoutes"`
	IsActive          bool     `json:"isActive"`
	FirstSeenAt       string   `json:"firstSeenAt"`
	ActivePeriodStart *string  `json:"activePeriodStart,omitempty"`
	ActivePeriodEnd   *string  `json:"activePeriodEnd,omitempty"`
	ResolvedAt        *string  `json:"resolvedAt,omitempty"`
}

// DelaySummary represents live delay statistics snapshot
type DelaySummary struct {
	TotalTrains     int     `json:"totalTrains"`
	DelayedTrains   int     `json:"delayedTrains"`
	OnTimePercent   float64 `json:"onTimePercent"`
	AvgDelaySeconds float64 `json:"avgDelaySeconds"`
	MaxDelaySeconds int     `json:"maxDelaySeconds"`
	WorstRoute      string  `json:"worstRoute,omitempty"`
}

// DelayHourlyStat represents hourly delay data for a route
type DelayHourlyStat struct {
	RouteID          string  `json:"routeId"`
	HourBucket       string  `json:"hourBucket"`
	ObservationCount int     `json:"observationCount"`
	MeanDelaySeconds float64 `json:"meanDelaySeconds"`
	StdDevSeconds    float64 `json:"stdDevSeconds"`
	OnTimePercent    float64 `json:"onTimePercent"`
	MaxDelaySeconds  int     `json:"maxDelaySeconds"`
}

// DelayStatsResponse is the response for GET /api/delays/stats
type DelayStatsResponse struct {
	Summary     DelaySummary      `json:"summary"`
	HourlyStats []DelayHourlyStat `json:"hourlyStats"`
	LastChecked time.Time         `json:"lastChecked"`
}

// AlertsResponse is the response for GET /api/alerts
type AlertsResponse struct {
	Alerts      []ServiceAlert `json:"alerts"`
	Count       int            `json:"count"`
	LastChecked time.Time      `json:"lastChecked"`
}

File: apps/web/src/lib/api/delays.ts
/**
 * Delays & Alerts API Client
 *
 * Fetches delay statistics and service alerts from the backend.
 */

import { fetchWithRetry } from './fetchWithRetry';

const API_BASE = import.meta.env.VITE_API_BASE || '/api';

// Types

export interface ServiceAlert {
  alertId: string;
  cause: string;
  effect: string;
  descriptionText: string;
  affectedRoutes: string[];
  isActive: boolean;
  firstSeenAt: string;
  activePeriodStart?: string;
  activePeriodEnd?: string;
  resolvedAt?: string;
}

export interface DelaySummary {
  totalTrains: number;
  delayedTrains: number;
  onTimePercent: number;
  avgDelaySeconds: number;
  maxDelaySeconds: number;
  worstRoute: string;
}

export interface DelayHourlyStat {
  routeId: string;
  hourBucket: string;
  observationCount: number;
  meanDelaySeconds: number;
  stdDevSeconds: number;
  onTimePercent: number;
  maxDelaySeconds: number;
}

export interface DelayStatsResponse {
  summary: DelaySummary;
  hourlyStats: DelayHourlyStat[];
  lastChecked: string;
}

export interface AlertsResponse {
  alerts: ServiceAlert[];
  count: number;
  lastChecked: string;
}

// API Functions

/**
 * Fetch delay statistics with optional route filter
 */
export async function fetchDelayStats(
  routeId?: string,
  period: string = '24h'
): Promise<DelayStatsResponse> {
  const params = new URLSearchParams({ period });
  if (routeId) params.set('route_id', routeId);

  const response = await fetchWithRetry(`${API_BASE}/delays/stats?${params}`, {
    logPrefix: 'Delays API',
    timeoutMs: 5000,
    useCircuitBreaker: false,
  });

  if (!response.ok) {
    throw new Error(`Failed to fetch delay stats: ${response.status}`);
  }

  return response.json();
}

/**
 * Fetch active service alerts with optional route filter
 */
export async function fetchAlerts(
  routeId?: string,
  lang?: string
): Promise<AlertsResponse> {
  const params = new URLSearchParams();
  if (routeId) params.set('route_id', routeId);
  if (lang) params.set('lang', lang);

  const url = params.toString()
    ? `${API_BASE}/alerts?${params}`
    : `${API_BASE}/alerts`;

  const response = await fetchWithRetry(url, {
    logPrefix: 'Alerts API',
    timeoutMs: 5000,
    useCircuitBreaker: false,
  });

  if (!response.ok) {
    throw new Error(`Failed to fetch alerts: ${response.status}`);
  }

  return response.json();
}
